federationclient.go

```diff
@@ -170,3 +171,30 @@ func (ac *FederationClient) LookupRoomAlias(
 	err = ac.doRequest(ctx, req, &res)
 	return
 }
+
+// Backfill asks a homeserver for events early enough for them to not be in the
+// local database.
+// See https://matrix.org/docs/spec/server_server/unstable.html#get-matrix-federation-v1-backfill-roomid
+func (ac *FederationClient) Backfill(
+	ctx context.Context, s ServerName, roomID string, limit int, eventIDs []string,
+) (res Transaction, err error) {
+	// Parse the limit into a string so that we can include it in the URL's query.
+	limitStr := strconv.Itoa(limit)
+
+	// Define the URL's query.
+	query := url.Values{}
+	query["v"] = eventIDs
+	query.Set("limit", limitStr)
+
+	// Use the url.URL structure to easily generate the request's URI (path?query).
+	u := url.URL{
+		Path:     "/_matrix/federation/v1/backfill/" + roomID,
```

Comment:
Does this not need a trailing slash? And to url encode the roomID?

Reply:
Does need a trailing slash indeed. roomID is automatically url encoded when calling `u.RequestURI()`, as well as the event IDs.

Reply:
Trailing slash fixed in https://github.com/matrix-org/gomatrixserverlib/pull/101/commits/6802ea9d837d411125f79c3a1e254022074726f3.

Reply:
Hmm. I didn't think of that. Fixed in 9682b8b.

Reply:
Fair, its one of the classic mistakes literally *everyone* makes :)